File: lab3_1/lab3_1/Car.cpp
#include "Car.h"

bool IsGearInSpeedFrames(Gear gear, int m_speed)
{
	if (gear == Gear::Reverse && m_speed == 0)
	{
		return true;
	}
	if (gear == Gear::Neutral)
	{
		return true;
	}
	if (gear == Gear::One && m_speed >= oneSpeedLimits.first && m_speed <= oneSpeedLimits.second)
	{
		return true;
	}
	if (gear == Gear::Two && m_speed >= twoSpeedLimits.first && m_speed <= twoSpeedLimits.second)
	{
		return true;
	}
	if (gear == Gear::Three && m_speed >= threeSpeedLimits.first && m_speed <= threeSpeedLimits.second)
	{
		return true;
	}
	if (gear == Gear::Four && m_speed >= fourSpeedLimits.first && m_speed <= fourSpeedLimits.second)
	{
		return true;
	}
	if (gear == Gear::Five && m_speed >= fiveSpeedLimits.first && m_speed <= fiveSpeedLimits.second)
	{
		return true;
	}

	return false;
}

//TODO: не понятная функция
bool Car::IsCorrectGearForThisSpeed(int in_speed, Direction direction)
{
	switch (m_gear)
	{
	case Gear::Neutral:
		switch (direction)
		{
		case Direction::Back:
			return (in_speed >= m_speed && in_speed <= 0);
		case Direction::Along:
			return (in_speed <= m_speed && in_speed >= 0);
		default:
			return false;
		}

	case Gear::Reverse:
		return (in_speed >= reverseSpeedLimits.first && in_speed <= reverseSpeedLimits.second);

	default:
		return (IsGearInSpeedFrames(m_gear, in_speed));
	}

	return false;
}
//дополнительные функции

bool Car::TurnOffEngine()
{
	if (m_isEngineTurnedOn && (m_gear == Gear::Neutral) && (m_speed == 0))
	{
		m_isEngineTurnedOn = false;
		return true;
	}
	return false;
}

bool Car::TurnOnEngine()
{
	if (!m_isEngineTurnedOn)
	{
		m_isEngineTurnedOn = true;
		return true;
	}
	return false;
}

bool Car::IsEngineTurnedOn() const
{
	return m_isEngineTurnedOn;
}


bool Car::SetGear(Gear gear)
{
	if (m_isEngineTurnedOn && IsGearInSpeedFrames(gear, m_speed))
	{
		m_gear = gear;
		return true;
	}
	return false;
}

Gear Car::GetGear() const
{
	return m_gear;
}

bool Car::SetSpeed(int moduleSpeed)
{
	int speed = moduleSpeed;
	//TODO: const negativeIndex лучше просто убрать +
	Direction direction = Car::GetDirection();

	if (speed < 0)
	{
		return false;
	}

	if (direction == Direction::Back || (direction == Direction::OnPlace && m_gear == Gear::Reverse))
	{
		speed = -speed;
	}

	if (m_isEngineTurnedOn && IsCorrectGearForThisSpeed(speed, direction))
	{
		m_speed = speed;
		return true;
	}
	return false;
}

//TODO: может вернуть отрицательно значение, так не должно быть +
int Car::GetSpeed() const
{
	if (m_speed < 0)
	{
		return -m_speed;
	}
	return m_speed;
}

Direction Car::GetDirection() const
{
	if (m_speed < 0)
	{
		return Direction::Back;
	}
	if (m_speed > 0)
	{
		return Direction::Along;
	}
	return Direction::OnPlace;
}

File: testing/lab3/TestCar/Car.cs
using System;

namespace CarSpace
{
    public enum Gear
    {
        Reverse = -1,
        Neutral = 0,
        One = 1,
        Two = 2,
        Three = 3,
        Four = 4,
        Five = 5
    };
    public enum Direction
    {
        Along,
        Back,
        OnPlace
    };
    public class Car
    {
        private int m_speed;
        private Gear m_gear;
        private string m_number;

        public Car(int speed, Gear gear, string number)
        {
            m_speed = speed;
            m_gear = gear;
            m_number = number;
        }

        (int, int) reverseSpeedLimits = (-20, 0);
        (int, int) oneSpeedLimits = (0, 30);
        (int, int) twoSpeedLimits = (20, 50);
        (int, int) threeSpeedLimits = (30, 60);
        (int, int) fourSpeedLimits = (40, 90);
        (int, int) fiveSpeedLimits = (50, 150);

        bool IsGearInSpeedFrames()
        {
            if (m_gear == Gear.Reverse && m_speed == 0)
            {
                return true;
            }
            if (m_gear == Gear.Neutral)
            {
                return true;
            }
            if (m_gear == Gear.One && m_speed >= oneSpeedLimits.Item1 && m_speed <= oneSpeedLimits.Item2)
            {
                return true;
            }
            if (m_gear == Gear.Two && m_speed >= twoSpeedLimits.Item1 && m_speed <= twoSpeedLimits.Item2)
            {
                return true;
            }
            if (m_gear == Gear.Three && m_speed >= threeSpeedLimits.Item1 && m_speed <= threeSpeedLimits.Item2)
            {
                return true;
            }
            if (m_gear == Gear.Four && m_speed >= fourSpeedLimits.Item1 && m_speed <= fourSpeedLimits.Item2)
            {
                return true;
            }
            if (m_gear == Gear.Five && m_speed >= fiveSpeedLimits.Item1 && m_speed <= fiveSpeedLimits.Item2)
            {
                return true;
            }

            return false;
        }

        public void SetCorrectGear()
        {
            Gear gear = ChooseCorrectGear();
            m_gear = gear;
        }

        Gear ChooseCorrectGear()
        {
            if (m_speed <= 0)
            {
                return Gear.Reverse;
            }
            if (m_speed >= oneSpeedLimits.Item1 && m_speed <= oneSpeedLimits.Item2)
            {
                return Gear.One;
            }
            if (m_speed >= twoSpeedLimits.Item1 && m_speed <= twoSpeedLimits.Item2)
            {
                return Gear.Two;
            }
            if (m_speed >= threeSpeedLimits.Item1 && m_speed <= threeSpeedLimits.Item2)
            {
                return Gear.Three;
            }
            if (m_speed >= fourSpeedLimits.Item1 && m_speed <= fourSpeedLimits.Item2)
            {
                return Gear.Four;
            }
            if (m_speed >= fiveSpeedLimits.Item1 && m_speed <= fiveSpeedLimits.Item2)
            {
                return Gear.Five;
            }
            return Gear.Neutral;
        }

        public Direction GetDirection()
        {
            if (m_speed < 0)
            {
                return Direction.Back;
            }
            if (m_speed > 0)
            {
                return Direction.Along;
            }
            return Direction.OnPlace;
        }

        public bool IsCorrectGearForThisSpeed(Direction direction)
        {
            switch (m_gear)
            {
                case Gear.Neutral:
                    switch (direction)
                    {
                        case Direction.Back:
                            return (m_speed <= 0);
                        case Direction.Along:
                            return (m_speed >= 0);
                        default:
                            return false;
                    }
                case Gear.Reverse:
                    return (m_speed >= reverseSpeedLimits.Item1 && m_speed <= reverseSpeedLimits.Item2);
                default:
                    return (IsGearInSpeedFrames());
            }
        }

        public string ToStr()
        {
            string isCorrectWork = "correct";
            if (IsCorrectGearForThisSpeed(GetDirection()) == false)
            {
                isCorrectWork = "incorrect";
            }
            return $"Car number:  ({m_number})  " +
                    $"Speed: ({m_speed})  " +
                    $"Gear: ({m_gear})  " +
                    $"Direction: {GetDirection()}  " +
                    $"It's work {isCorrectWork}";
        }
    }
}